src/metacognition/failure_predictor.py

```
@@ -0,0 +1,451 @@
+    - Action prerequisites
+    - Recent failure patterns
+    """
+
+    # Thresholds for prediction
+    HIGH_FAILURE_THRESHOLD = 0.7
+    MEDIUM_FAILURE_THRESHOLD = 0.4
+    MIN_OBSERVATIONS_FOR_PREDICTION = 3
+    RECENT_WINDOW_HOURS = 24
+
+    # Rate limit tracking (action -> (count, reset_time))
+    _rate_limits: Dict[str, tuple[int, datetime]] = {}
```

Ruff (0.14.10)

Mutable class attributes should be annotated with `typing.ClassVar`

(RUF012)